File: app/src/main/java/com/vmenon/mpo/core/persistence/EpisodeDao.java
package com.vmenon.mpo.core.persistence;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.Query;

import com.vmenon.mpo.api.Episode;

import java.util.List;

import static androidx.room.OnConflictStrategy.REPLACE;

@Dao
public interface EpisodeDao {
    @Insert(onConflict = REPLACE)
    void save(Episode episode);

    @Query("SELECT * FROM episode")
    LiveData<List<Episode>> load();

    @Query("SELECT * from episode WHERE id = :id")
    Episode byId(long id);

    @Query("SELECT * from episode WHERE id = :id")
    LiveData<Episode> liveById(long id);
}

File: app/src/main/java/com/vmenon/mpo/core/persistence/EpisodeDao.kt
package com.vmenon.mpo.core.persistence

import androidx.room.Dao
import androidx.room.Query

import com.vmenon.mpo.model.EpisodeModel
import com.vmenon.mpo.model.EpisodeWithShowDetailsModel
import io.reactivex.Flowable

@Dao
abstract class EpisodeDao : BaseDao<EpisodeModel> {
    @Query(
        """
        SELECT * from episode 
        INNER JOIN show on episode.showId = show.id
        """
    )
    abstract fun getAllWithShowDetails(): Flowable<List<EpisodeWithShowDetailsModel>>

    @Query(
        """
        SELECT * from episode 
        INNER JOIN show on episode.showId = show.id
        WHERE episode.id = :id
        """
    )
    abstract fun getWithShowDetailsById(id: Long): Flowable<EpisodeWithShowDetailsModel>
}
